### DISCUSSION TITLE
Backup Process

### DISCUSSION THREAD

**Discussion by TheNore** [jasonacox/Powerwall-Dashboard#95] [Q&A] [answered]
Hello, I tried usign the backup script but the process halts when it tires to tar the backup files - from what it seems, there are no files populated into the /backups folder. 

> ``:~/Powerwall-Dashboard/backups $ sudo ./backup.sh
> Creating InfluxDB Backup
> 2022/10/03 11:59:42 backing up metastore to /var/lib/influxdb/backups/meta.00
> 2022/10/03 11:59:42 backing up db=powerwall
> 2022/10/03 11:59:42 backing up db=powerwall rp=autogen shard=3 to /var/lib/influxdb/backups/powerwall.autogen.00003.00            since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:42 backing up db=powerwall rp=autogen shard=17 to /var/lib/influxdb/backups/powerwall.autogen.00017.0           0 since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:42 backing up db=powerwall rp=autogen shard=38 to /var/lib/influxdb/backups/powerwall.autogen.00038.0           0 since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:43 backing up db=powerwall rp=raw shard=32 to /var/lib/influxdb/backups/powerwall.raw.00032.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:43 backing up db=powerwall rp=raw shard=34 to /var/lib/influxdb/backups/powerwall.raw.00034.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:43 backing up db=powerwall rp=raw shard=36 to /var/lib/influxdb/backups/powerwall.raw.00036.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:43 backing up db=powerwall rp=raw shard=39 to /var/lib/influxdb/backups/powerwall.raw.00039.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pwtemps shard=5 to /var/lib/influxdb/backups/powerwall.pwtemps.00005.00            since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pwtemps shard=19 to /var/lib/influxdb/backups/powerwall.pwtemps.00019.0           0 since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pwtemps shard=40 to /var/lib/influxdb/backups/powerwall.pwtemps.00040.0           0 since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=vitals shard=6 to /var/lib/influxdb/backups/powerwall.vitals.00006.00 s           ince 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=vitals shard=20 to /var/lib/influxdb/backups/powerwall.vitals.00020.00            since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=vitals shard=41 to /var/lib/influxdb/backups/powerwall.vitals.00041.00            since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=kwh shard=4 to /var/lib/influxdb/backups/powerwall.kwh.00004.00 since 0           001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=kwh shard=21 to /var/lib/influxdb/backups/powerwall.kwh.00021.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=kwh shard=42 to /var/lib/influxdb/backups/powerwall.kwh.00042.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=daily shard=9 to /var/lib/influxdb/backups/powerwall.daily.00009.00 sin           ce 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=daily shard=22 to /var/lib/influxdb/backups/powerwall.daily.00022.00 si           nce 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=daily shard=43 to /var/lib/influxdb/backups/powerwall.daily.00043.00 si           nce 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=monthly shard=10 to /var/lib/influxdb/backups/powerwall.monthly.00010.0           0 since 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=grid shard=7 to /var/lib/influxdb/backups/powerwall.grid.00007.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=grid shard=23 to /var/lib/influxdb/backups/powerwall.grid.00023.00 sinc           e 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=grid shard=44 to /var/lib/influxdb/backups/powerwall.grid.00044.00 sinc           e 0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pod shard=8 to /var/lib/influxdb/backups/powerwall.pod.00008.00 since 0           001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pod shard=24 to /var/lib/influxdb/backups/powerwall.pod.00024.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backing up db=powerwall rp=pod shard=45 to /var/lib/influxdb/backups/powerwall.pod.00045.00 since            0001-01-01T00:00:00Z
> 2022/10/03 11:59:44 backup complete:
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/meta.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.autogen.00003.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.autogen.00017.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.autogen.00038.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.raw.00032.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.raw.00034.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.raw.00036.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.raw.00039.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pwtemps.00005.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pwtemps.00019.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pwtemps.00040.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.vitals.00006.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.vitals.00020.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.vitals.00041.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.kwh.00004.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.kwh.00021.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.kwh.00042.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.daily.00009.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.daily.00022.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.daily.00043.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.monthly.00010.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.grid.00007.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.grid.00023.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.grid.00044.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pod.00008.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pod.00024.00
> 2022/10/03 11:59:44     /var/lib/influxdb/backups/var/lib/influxdb/backups/powerwall.pod.00045.00
> Backing up Powerwall-Dashboard (influxdb grafana)
> tar (child): /home/pi/Powerwall-Dashboard/backup/Powerwall-Dashboard.2022-10-03.tgz: Cannot open: No such file or dire           ctory
> tar (child): Error is not recoverable: exiting now
> influxdb/
> influxdb/wal/
> influxdb/wal/_internal/
> influxdb/wal/_internal/monitor/
> influxdb/wal/_internal/monitor/33/
> influxdb/wal/_internal/monitor/33/_00034.wal
> influxdb/wal/_internal/monitor/27/
> influxdb/wal/_internal/monitor/35/
> influxdb/wal/_internal/monitor/35/_00035.wal
> influxdb/wal/_internal/monitor/29/
> influxdb/wal/_internal/monitor/37/
> influxdb/wal/_internal/monitor/37/_00035.wal
> influxdb/wal/_internal/monitor/31/
> influxdb/wal/_internal/monitor/25/
> influxdb/wal/_internal/monitor/46/
> influxdb/wal/_internal/monitor/46/_00021.wal
> tar: /home/pi/Powerwall-Dashboard/backup/Powerwall-Dashboard.2022-10-03.tgz: Cannot write: Broken pipe
> tar: Child returned status 2
> tar: Error is not recoverable: exiting now
> Cleaning up old backups
> find: ‘/home/pi/Powerwall-Dashboard/backup/Powerwall-Dashboard.*tgz’: No such file or directory
> Done
> ``

**Comment by jasonacox** [comment, marked as answer]
Hi @TheNore  - thanks for opening this issue!

It is having trouble writing the tar file to `/home/pi/Powerwall-Dashboard/backup/` - There is a typo in the `BACKUP_FOLDER` variable you are using.  It should be `backups` with an `s`.  You should be able to make the edits in the backup.sh file:

```bash
BACKUP_FOLDER="/home/user/Powerwall-Dashboard/backups"    # Destination folder for backups
DASHBOARD="/home/user/Powerwall-Dashboard"                # Location of Dashboard to backup
```

I notice that the backup.sh file has it as "/backup" by default which is probably why you did the same.  I'm going to adjust it to help others.

**Reply by oralallen82** [reply]
> Hi @TheNore - thanks for opening this issue!
> 
> It is having trouble writing the tar file to `/home/pi/Powerwall-Dashboard/backup/` - There is a typo in the `BACKUP_FOLDER` variable you are using. It should be `backups` with an `s`. You should be able to make the edits in the backup.sh file:
> 
> ```shell
> BACKUP_FOLDER="/home/user/Powerwall-Dashboard/backups"    # Destination folder for backups
> DASHBOARD="/home/user/Powerwall-Dashboard"                # Location of Dashboard to backup
> ```
> 
> I notice that the backup.sh file has it as "/backup" by default which is probably why you did the same. I'm going to adjust it to help others.

@jasonacox I noticed the deletion no longer works? Did something change from the last version? the last version worked fine.

**Reply by jasonacox** [reply]
Hi @oralallen82 - the change should not have impacted the deletion part, but I suspect the directories may have reverted if you did an upgrade.  The changes made: https://github.com/jasonacox/Powerwall-Dashboard/commit/725bf236d6ed5d9da34dadd1d7665e1f207b6ac6

Did you backup.sh file revert?

**Reply by oralallen82** [reply]
I re-ran an upgrade and modified the file to reflect the correct parameters. It appears to be working now. Thanks as always @jasonacox

**Reply by jasonacox** [reply]
Awesome!  My only TODO on this is that upgrade process will keep replacing this file and remove the localized (directory path) updates you make.  Thoughts or suggestions?

**Reply by mcbirse** [reply]
Why not create a `backup.conf` file where the user configurable environment variables are set?

i.e. backup.conf to contain:
```sh
# Set values for your environment 
DASHBOARD="/home/user/Powerwall-Dashboard" # Location of Dashboard to backup
BACKUP_FOLDER="${DASHBOARD}/backups"          # Destination folder for backups
KEEP="5"   
```

Then update the script to source the config defined environment variables.

```sh
. backup.conf
```

**Reply by jasonacox** [reply]
I thought about that but two issues:

1) If we had a template backup.conf file, it would be overwritten by upgrade process too.
2) It would be an issue if a person set this to be run by a cronjob or just ran it outside the directory with the backup.conf file (which could be fixed by some test logic or discover current path logic)

What if we just renamed it to `backup.example` with instruction to copy it to `backup.sh`?  

Alternatively, the script could be adjust to require both variables interactively or be set on command line:

```bash
backup.sh /home/user/Powerwall-Dashboard /home/user/Powerwall-Dashboard/backups
```

If interactively, the first time you run it, it could ask for those variable and store them in a created backup.conf file for you (not overwritten by upgrade).  The other logic for the "current directory" could be added - something like:

```bash
MY_PATH=$(dirname "$0")            # relative path
MY_PATH=$(cd "$MY_PATH" && pwd)    # absolute path
if [[ -z "$MY_PATH/backup.conf" ]] ; then
  echo "Setup: missing config file - Please set backup locations:"
  # TODO
fi
source "$MY_PATH/backup.conf" 
# rest of backup script...
```

**Reply by mcbirse** [reply]
> If interactively, the first time you run it, it could ask for those variable and store them in a created backup.conf file for you (not overwritten by upgrade).

I like this idea... in fact was just grappling with this problem for the Tesla history tool to make it as user friendly (and Git friendly) as possible. I ended up doing the same so it will interactively prompt the user for the config on first run and save that to a file - therefore the config file won't be tracked/affected by Git. Seems like a good solution. 👍

**Comment by oralallen82** [comment]
Is there a way to not upgrade a specific file if the version for that file
doesn't change? So let's say backup.sh file version is 2.6.2 the next
update you make you didn't modify the backup.sh everything else would get
upgraded except that?

Or during the upgrade; unless you upgraded the backup.sh file, the upgrade
process stores the contents of the file then restores it on completing?

**Reply by jasonacox** [reply]
Hi @oralallen82 , good question.  I have been thinking about updating the upgrade process. The current method just uses git to pull the files and forces overwrite merge, specifically:

```
git stash
git pull --rebase
```

Unfortunately, that will always wipe out local files with files in the repo, but doesn't touch additional files which is why env files are safe.

A better method is probably using the release archive (zip) or better yet, an upgrade bundle to install what was changed (to your point). I need to do more research to see what else is available. I'm sure others have solved this.  If anyone has any time and expertise in this area, I would love a volunteer to help build a better upgrade system. 🙏
